### PR TITLE
Contracts and Harnesses for `<*mut T>::add`, `sub` and `offset`

### PR DESCRIPTION
**Pull request by stogaru**
Towards #76 

### Changes
* Adds contracts for `<*mut T>::add`, `<*mut T>::sub` and `<*mut T>::offset`
* Adds proofs for contracts of the above functions verifying the pointee types:
  * All integer types
  * Tuples (composite type)
  * Unit Type
* Defines a macro for add and sub and another for offset.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/core/src/ptr/mut_ptr.rs [hunk 1]
```diff
@@ -403,6 +407,10 @@ impl<T: ?Sized> *mut T {
     #[rustc_const_stable(feature = "const_ptr_offset", since = "1.61.0")]
     #[inline(always)]
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+    #[requires(kani::mem::can_dereference(self))]
+    // TODO: Determine the valid value range for 'count' and update the precondition accordingly.
```
Comment (zhassan-aws) [inline library/core/src/ptr/mut_ptr.rs:411]: The precondition on `count` should be similar to what is used in https://github.com/model-checking/verify-rust-std/pull/105

Comment (stogaru) [reply library/core/src/ptr/mut_ptr.rs:411]: Is it ok if we have a workaround for the precondition in the harness similar to what is used in the [same PR](https://github.com/model-checking/verify-rust-std/pull/105/files#diff-430db13219f4c1ac4f4d35f685a45778572e039dab837c5ffeec9a810fc547b4R1858)?

Comment (zhassan-aws) [reply library/core/src/ptr/mut_ptr.rs:411]: You can use the same precondition for testing. But before the PR is merged, we need the implementation to be based on the new API introduced in https://github.com/model-checking/kani/pull/3583.

Comment (stogaru) [reply library/core/src/ptr/mut_ptr.rs:411]: We went through the new API which allows us to prove if two pointers are pointing to the same allocation. Are you expecting a precondition using that API on count? Would a precondition like below work?

```rust
#[requires(
            count.checked_mul(core::mem::size_of::<T>()).is_some() &&
            (self as isize).checked_add(count * core::mem::size_of::<T>() as isize).is_some() && 
            same_allocation( self, ((self as isize) + (count * core::mem::size_of::<T>() as isize)) as *mut T
)]
```

**Comment by xsxszab** [comment]
@zhassan-aws Hi the proofs have been updated to use pointer generator, however after the update we encountered a strange test failure, probably a bug: the test case `check_mut_add_u8` constantly fails with the following output:
```
SUMMARY:
 ** 1 of 293 failed (4 unreachable)
Failed Checks: Only a single top-level call to function _RNCNCNvMNtNtCs1YNGORL5zck_4core3ptr7mut_ptrOh3adds0_0s_0Ba_ when checking contract _RNCNCNvMNtNtCs1YNGORL5zck_4core3ptr7mut_ptrOh3adds0_0s_0Ba_
 File: "/Users/xsxsz/vscode-projects/verify-rust-std/library/core/src/ptr/mut_ptr.rs", line 1019, in _RNCNCNvMNtNtCs1YNGORL5zck_4core3ptr7mut_ptrOh3adds0_0s_0Ba_

VERIFICATION:- FAILED
Verification Time: 0.45757997s
```
All other proofs generated by the same macro work fine. Is there any possible cause for this problem? For now, we have to skip this test case by commenting it out.

Additionally, we noticed a performance overhead when using the pointer generator: The compile time for verifying the standard library increased significantly from around 40 seconds to 1m40s. It would be great to see optimizations for this in the future.

**Comment by zhassan-aws** [comment]
@celinval encountered the same issue in https://github.com/model-checking/verify-rust-std/pull/37#issuecomment-2455780856. @celinval were you able to figure out what is causing the error and whether there's a workaround?

**Comment by celinval** [comment]
Yes, the problem is that the pointer generator code is likely using the API you are trying to verify. For the harness that is failing, you won't be able to use the generator as is today. Sorry!

**Review by feliperodri — APPROVED** [review]

File: library/core/src/ptr/mut_ptr.rs [hunk 2]
```diff
@@ -400,6 +404,20 @@ impl<T: ?Sized> *mut T {
...
+    #[requires(
+        // Precondition 1: the computed offset `count * size_of::<T>()` does not overflow `isize`
+        count.checked_mul(core::mem::size_of::<T>() as isize).is_some() &&
+        // Precondition 2: adding the computed offset to `self` does not cause overflow
+        (self as isize).checked_add((count * core::mem::size_of::<T>() as isize)).is_some() &&
+        // Precondition 3: If `T` is a unit type (`size_of::<T>() == 0`), this check is unnecessary as it has no allocated memory.
+        // Otherwise, for non-unit types, `self` and `self.wrapping_offset(count)` should point to the same allocated object,
+        // restricting `count` to prevent crossing allocation boundaries.
+        ((core::mem::size_of::<T>() == 0) || (kani::mem::same_allocation(self, self.wrapping_offset(count))))
+    )]
```
[the remark this answers is not in the document]
Comment (xsxszab) [reply library/core/src/ptr/mut_ptr.rs:418]: Addressed in commit b528b0e. Apologies for the delay; I thought I would receive notifications on this PR, but it seems that wasn't the case.

File: library/core/src/ptr/mut_ptr.rs [hunk 3]
```diff
@@ -2302,3 +2350,125 @@ impl<T: ?Sized> PartialOrd for *mut T {
...
+
+    // <*mut T>:: add() integer types verification
+    generate_mut_arithmetic_harness!(i8, check_mut_add_i8, add);
+    generate_mut_arithmetic_harness!(i16, check_mut_add_i16, add);
+    generate_mut_arithmetic_harness!(i32, check_mut_add_i32, add);
+    generate_mut_arithmetic_harness!(i64, check_mut_add_i64, add);
+    generate_mut_arithmetic_harness!(i128, check_mut_add_i128, add);
+    generate_mut_arithmetic_harness!(isize, check_mut_add_isize, add);
+    // Due to a bug of kani this test case is malfunctioning for now
+    // generate_mut_arithmetic_harness!(u8, check_mut_add_u8, add);
```
[the remark this answers is not in the document]
Comment (xsxszab) [reply library/core/src/ptr/mut_ptr.rs:2419]: Sure, the tracking issue is [here](https://github.com/model-checking/kani/issues/3743).

**Review by feliperodri — APPROVED** [review]

**Review by tautschnig — APPROVED** [review]

**Pull request merged**